### ISSUE TITLE
Introduce `scrollPositionRestoration` for `NavigationExtras`

### ISSUE THREAD

**Issue by piernik** [angular/angular#26744]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
In `angular 6` You introduced great new feature `scrollPositionRestoration` in router.
When it is enabled it **always** scroll to top on navigation change.
But sometimes we don't need it (example with nested router-outlets). 

How about introducing new `NavigationExtras` param: `scrollPositionRestoration`.
It will enable or disable default behaviour only for that navigation change.

**Comment by borriej** [comment]
I need and love this feature as well!

E.g.

I have globally:
```
@NgModule({
  imports: [RouterModule.forRoot(routes, {
    scrollPositionRestoration: 'enabled',
    anchorScrolling: 'enabled'
  })],
  exports: [RouterModule]
})
export class AppRoutingModule { }
```

On a overview page i have queryParams that update when I add filters...
```
  updateUrl(): void {
    const options: NavigationExtras = {
      replaceUrl: true,
      relativeTo: this.route,
      queryParams: { date: this.dateValue(), upperNode: this.upperNodeValue() }
    };

    this.router.navigate([], options);
  }
```

Now the page scrolls up, but on this overview page... i don't want to scroll to the top. 
So an extra prop on ```NavigationExtra``` to overwrite this behavior would be great

**Comment by pnm-i** [comment]
Hi borriej, same problem.
As temporary solution I use:
`const urlTree = this.router.createUrlTree([], {
          queryParams: filterModel
        });
        this.location.replaceState(urlTree.toString())`
or
`this.location.go(urlTree.toString())`

**Comment by hohler** [comment]
The workaround from @pnm-i does not trigger the `this.activatedRoute.queryParamMap.subscribe()` observable..
We need that feature!

**Comment by FDIM** [comment]
Apparently we also need this feature. Default enabled strategy works for 95% of the cases and it is great!
On the flip side we have a single page where we load more items on scroll, and when new item covers most screen estate we update the url to reflect that - now in this case we do not want router to affect scroll position and would like to have a way to prevent it. I thought that replaceUrl: true option would do that but apparently not. Maybe it could be a good default?

Anyway, I'll investigate and see if I can make a PR for this soonish.

**Comment by iffa** [comment]
I agree, this would be nice to have. We store some user selections in the route & on long pages scrolling to the top when user makes a selection is not a very good user experience.

**Comment by webcourse-ng** [comment]
Trying to implement @pnm-i's workaround, but getting a 'cannot find name' error on filterModel. I've searched the net for a module that contains it, but am not seeing much on it. Someone please nudge me in the right direction.

**Comment by julmot** [comment]
Urgently need this too.

**Comment by julmot** [comment]
Urgently need this too.

Like all the other use cases here, by default I want to scroll to [0, 0] on route change. But there's one issue that I'm facing: When having filter query params, I don't want to scroll to top all the time. Instead, in this scenario, I want to keep the scroll position. Saving it using `const scrollPosition = this.viewportScroller.getScrollPosition()` and later restoring it manually `this.viewportScroller.scrollToPosition(scrollPosition)` will cause a flicker effect, since Ng will scroll to top, afterwards I will scroll to the previous position.

**Comment by MehZmeH** [comment]
+1

Would like to use this feature for disabling scrollPositionRestoration for a single router.navigate(...) call.

**Comment by atscott** [comment]
Copying information from #38382

When scrollPositionRestoration is enabled the behaviour on forward navigation is always to go to viewport [0,0]. This may be undesirable when auxilary routes are implemented but it is not configurable or overrideable at either an outlet or a route.

## 🔬 Minimal Reproduction
https://auxiliary-route-modal-animated-basic-working-example.stackblitz.io

(Scroll down to the link to navigate to an auxilary route. In this example the route might open a modal that gives context to the primary outlet and it's scrolled position - therefore it is undesirable to scroll away from the link. But with scrollPositionRestoration enabled (often desired on the primary router outlet) it will take you back to the top.

**Comment by anbaran** [comment]
> I agree, this would be nice to have. We store some user selections in the route & on long pages scrolling to the top when user makes a selection is not a very good user experience.
> 
> For anyone else facing a similar issue, here's a Gist with custom logic, that ignores query parameter changes as "forward navigation" (which one would argue does not actually count as forward navigation): https://gist.github.com/iffa/9c820072135d25a6372d58075fe264dd

It worked for me, Thanks a lot. I also added note in your gist to use: `this.viewportScroller.setHistoryScrollRestoration('manual');`

**Comment by kepek** [comment]
👍

**Comment by HinnHabib** [comment]
👍

**Comment by MatthewLHolden** [comment]
Was this ever implemented?

**Comment by JWess** [comment]
This seems like a great idea. I am here because I also am in need of disabling the scroll-to-top when query params change.

**Comment by mmazloum** [comment]
Wonder if the problem has been solved since but the easiest workarround if found is to use a non-existing anchor in your navigationextras. something like this.router.navigate([...], {fragment: 'youranchorthatdoesnotexist'}) with       anchorScrolling: 'enabled' in your Root RouterModule

**Comment by htrex** [comment]
@mmazloum workaround works, but a proper option  for NavigationExtras would be much appreciated for special cases like infinite scroll pages with the page parameter change reflected in the URL.

**Comment by hisham** [comment]
Thanks @mmazloum for the workaround. It works. But +1 for angular team to support a more official solution.

**Comment by YuryHowru** [comment]
Another workaround, which allowed me to keep `withInMemoryScrolling({ scrollPositionRestoration: 'enabled' })`

In the component, where I want to change queryParam without triggering the scroll:
`
location = inject(Location);
queryParamChangeWithoutScroll() {
  this.location.replaceState(currentRoute, `queryParamName=queryParamValue`);
}
`

This does not trigger any event on the router or active route, so have to store and update the current queryParam imperatively in a variable.

**Comment by stefanjudis** [comment]
Here's another workaround. We wanted to keep the scroll position whenever a `?sidebar` query param was added. And to do it, we changed `scrollPositionRestoration` and made it a dynamic getter.

```
RouterModule.forRoot(routes, {
  get scrollPositionRestoration() {
    const params = new URLSearchParams(window.location.search);
    if (params.get('sidebar')) {
      debugger;
      return 'disabled' as const;
    }
    return 'enabled' as const;
  },
  scrollOffset: [0, 164],
  anchorScrolling: 'enabled',
}),
```

This way, anytime Angular checks `this.options.scrollPositionRestoration` we can dynamically check if we want to enable scroll position restoration or not. :)

**Comment by naveedahmed1** [comment]
@stefanjudis should this approach work with standalone apps as well? I tried the below code but sometimes it work and sometimes it doesnt work:

```
provideRouter(APP_ROUTES,
  withEnabledBlockingInitialNavigation(),
  withInMemoryScrolling({
    get scrollPositionRestoration() {
      console.log('window.location.pathname', window.location.pathname);
      if (window.location.pathname.startsWith('/path/to/mycomponent)) {
        return 'disabled' as const;
      } else {
        return 'enabled' as const;
      }
    }
  }),

  withViewTransitions({ skipInitialTransition: true }),
  withComponentInputBinding()
),
```
